### ISSUE TITLE
[suggestion] eBPF/XDP transport for much faster throughput and less latency

### ISSUE THREAD

**Issue by takaomag** [nanomsg/nng#1301]
I am writing a server software and planning to use nng as a default transport stack between its components.
I guess eBPF/XDP(AF_XDP) transport will improve the throughput/latency especially on 10Gbps or more link, though it requires user-land TCP/IP stack library (like [smoltcp](https://github.com/smoltcp-rs/smoltcp) in case of Rust. I don't know a similar C library).

Is it bad idea? It would be nice if there is a plan to implement such transport.

**Comment by gdamore** [comment]
Running user space tcp is usually misguided. For very specific cases involving high frequency trading where every nanosecond of latency matters it can be useful but that is a tiny minority of applications.

I think using bpf for nng is unrealistic as well.  It would instead be preferable to move NNG into a kernel module (driver) along with your application.  This of course carries a lot of risk and your application will be non portable and a lot harder to manage (at least for most devs).

If your concern is throughput in terms of bytes per second you should just use larger message sizes.  You should be able to get to 25Gbps that way pretty easily.

Small packets are a different problem. Those will take a lot work to get up there.  And that is irrespective of kernel or user space - tiny frames maximize the work done per frame.

Get Outlook<https://aka.ms/qtex0l> for iOS
________________________________

**Comment by takaomag** [comment]
Yes, my main concern is throughput. I understand it, and my idea seems to be a premature optimization.
I will think and test deeply on nng and another solutions.
Thanks.
